### PR TITLE
Use JSON API and capture new bike_list data

### PR DESCRIPTION
**Pull request by jlogan540**
## A rough outline of the improvements are:
- Update API URL to grab data as JSON as this can be handled easily with .j.k. Removed xml specific code as a result.
- Add basic tests of the JSON parsing functionality.
- replaced wget with curl and have the output directed to stdout instead of using the file bikes.xml to store it.
- params seemed to be identical to .proc.params so I updated all references to it.
- webpage flag seems unnecessary as the API URL will probably only change infrequently, moved this into config instead.
- Use inbuilt TorQ logging for fullbikedataprotected instead of show.
- Capture new bike_list data coming from the API.
## May also address:
- [x]  Add namespace for bikes code.
- [ ] Having date passed to writedown function and removing all dates <.z.D-1 instead of =.z.D-2.
- [x] Using .Q.hg instead of curl.
- [ ] Move code/belfastbikes/bikes.q to code/processes/belfastbikes.q.
- [ ]  Update README.md to include minimum working version of kdb.
- [ ] Creating a release tag after merge.

### CODE REVIEW

File: code/belfastbikes/bikes.q [hunk 1]
```diff
@@ -1,54 +1,117 @@
...
+  bike_tab:update name: (exec raze (count each bike_list) #' enlist each name from tab) from exec raze bike_list from tab;
+  /Refactor data and extract relevant data
+  tab:`address`bike_list`spot`bike_types`bike _`time xcols update time:.z.P^t,name:trim name from tab;
+  bike_tab:`pedelac_battery`battery_pack _ `time`name xcols update time:.z.P^t,name:trim name,number:"I"$number,lock_types:raze lock_types from bike_tab;
+  /Convert floats to ints where appropriate
+  tab:@[tab;`uid`number`bikes`bike_racks`free_racks;`int$];
+  tab:@[tab;`place_type`bike_numbers;"I"$];
+  bike_tab:@[bike_tab;`bike_type`boardcomputer;`int$];
+  /Convert data to symbols from strings
+  //bike_tab:@[bike_tab;`name`state;`$];
```
Comment (tsmyth-aquaq) [inline code/belfastbikes/bikes.q]: Remove this comment.

File: README.md [hunk 2]
```diff
@@ -44,28 +58,49 @@ This project requires the use of KDB+ and the [TorQ framework](https://github.co
...
 ```
 ~/deploy$ ./start_bikes.sh
 ```
 This launches the bikes.q script wrapped in the TorQ framework.
 
 ## Collecting Data
-The process will run for 14 day once it starts, collecting data every 30 seconds. During the 14 days, there will be a write down to hdb at 6am every day using the previous day's data and saved by date. Within the bikes.q script there are timer functions for both the collection of data and the writedown which can be modified.
 
-The xmllogs directory will contain previously collected data in its raw XML format for each day, saved as a plain text file. 
+The process will run for 14 day once it starts, collecting data every 30 seconds.
```
Comment (jemmaborland) [inline README.md]:
```suggestion
The process will run for 14 days once it starts, collecting data every 30 seconds.
```

File: README.md [hunk 3]
```diff
@@ -44,28 +58,49 @@ This project requires the use of KDB+ and the [TorQ framework](https://github.co
...
 
 	~/deploy$ ls
-        appconfig  aquaq-torq-brochure.pdf  bikes.xml  code  config  docs  hdb  html  lib  LICENSE  logs  mkdocs.yml  README.md  setenv.sh  start_bikes.sh  tests  torq.q  xmllogs     
+        appconfig  aquaq-torq-brochure.pdf  code  config  docs  hdb  html  lib  LICENSE  logs  mkdocs.yml  README.md  setenv.sh  start_bikes.sh  tests  torq.q  jsonlogs     
 	
 ## Configuration
 
-You can change the city that you want to collect data for by changing the city number variable in `deploy/setenv.sh`, the default value is 238 (Belfast).
+You can change the city that you want to collect data for by changing the city
+number variable in `deploy/setenv.sh`, the default value is 238 (Belfast).
```
Comment (jemmaborland) [inline README.md]:
```suggestion
number variable in `deploy/setenv.sh`; the default value is 238 (Belfast).
```

File: README.md [hunk 4]
```diff
@@ -1,22 +1,36 @@
...
+database to perform analysis on, made possible by the TorQ framework. Taking
+advantage of TorQ's capabilities, the process has been enhanced to include error
+logging, history logging and an extension of the kdb+ built-in timer, making it
+repeat until a specified end-time. This allows us to view log files
+for any date and rebuild the usage of BelfastBikes for any date where we have
+collected information.
+
+Once a database has been created, time-series and other queries can be executed
+against the data.  For example, by rebuilding the table for yesterday's bike
+usage, we can create a mapping of 'routes' that each bike has taken, I.e. where
```
Comment (jemmaborland) [inline README.md]:
```suggestion
usage, we can create a mapping of 'routes' that each bike has taken, i.e. where
```

File: README.md [hunk 5]
```diff
@@ -44,28 +58,49 @@ This project requires the use of KDB+ and the [TorQ framework](https://github.co
...
+The jsonlogs directory will contain previously collected data in its raw JSON 
+format for each day, saved as a plain text file. 
+
+## Replaying Logs
+
+In order to replay a log file on disk, the following can be used:
+```
+.bikes.replayjsonlog 2019.01.01
+```
+This reads the data to the in memory table `place`. Which can then be written to
```
Comment (jemmaborland) [inline README.md]:
```suggestion
This reads the data to the in-memory table `place`, which can then be written to
```

File: README.md [hunk 6]
```diff
@@ -44,28 +58,49 @@ This project requires the use of KDB+ and the [TorQ framework](https://github.co
...
+disk with:
+```
+.bikes.writedown 2019.01.01
+```
 
 ## Example Usage
 
-To query the persisted data in the hdb, we can either load in a specific date partition to a q session, or load the entire database to perform queries across a range of dates. To load in the hdb to a q session we can run the following command:
+To query the persisted data in the hdb, we can either load in a specific date
```
Comment (jemmaborland) [inline README.md]:
```suggestion
To query the persisted data in the HDB, we can either load in a specific date
```

File: README.md [hunk 7]
```diff
@@ -44,28 +58,49 @@ This project requires the use of KDB+ and the [TorQ framework](https://github.co
...
+partition to a q session, or load the entire database to perform queries across
+a range of dates. To load in the hdb to a q session we can run the following command:
```
Comment (jemmaborland) [inline README.md]:
```suggestion
a range of dates. To load in the HDB to a q session we can run the following command:
```

**Review by tsmyth-aquaq — APPROVED** [review]

**Review by jemmaborland — APPROVED** [review]
Actual code looks good to me, just minor tweaks in readme

**Review by picoDoc — CHANGES_REQUESTED** [review]
great work 👍  Tested it out and all seems to work for me.  Two additions/issues that I found:
* can you add a stop script that cleanly shuts it down, saving data to disk?
* I noticed that the writedown will overwrite whatever is currently in the hdb for that date, so for example if you shutdown midday, and started back up, the next save down would overwrite the data before the shutdown.  Can you fix this?

File: README.md [hunk 8]
```diff
@@ -105,4 +142,3 @@ date      | x   Dock
 2017.09.24| 127 "Donegall Quay"
 2017.09.25| 41  "Alfred Street / St Malachy's Church"
 ```
-
```
Comment (jemmaborland) [inline README.md:108]: Can you add a bit in here about the usage of stop_bikes.sh?

File: code/util/intradaybikeswd.q [hunk 9]
```diff
@@ -0,0 +1,6 @@
+// Gets the rdb port past in from the command line.
+conn:.Q.def[(enlist `conn)!enlist 1;.Q.opt .z.x][`conn]
```
Comment (jemmaborland) [inline code/util/intradaybikeswd.q]: is 1 a sensible default here? would it worthwhile making it a null, then error trapping below if the port isn't open/null/can't be connected to for whatever reason?

**Review by picoDoc — APPROVED** [review]

**Review by jemmaborland — APPROVED** [review]

**Pull request merged**